[Client] Add Events for efficient cross-thread notification

batiati opened the pull request:
Removes the local TCP server hack being used in `tb_client.Signal` to facilitate cross-thread notification.

Instead, it uses the most efficient mechanism for waking the IO abstraction: `eventfd` on Linux, `EVFILT_USER` on macOS, and `PostQueuedCompletionStatus` on Windows.

Reviewed from previous work https://github.com/tigerbeetle/tigerbeetle/pull/622

matklad commented on src/io/linux.zig:
```diff
@@ -1301,6 +1301,89 @@ pub const IO = struct {
+        assert(bytes == @sizeOf(u64));
+    }
+
+    pub fn close_event(self: *IO, event_fd: posix.fd_t, completion: *Completion) void {
```
Let's not pass `Completion`? We don't use it, and we don't pass it in close_socket either.

matklad replied:
Hm, what is the cancelation story here? At the moment we call `close`, there might be an outstanding `event_listen`. It might even have fired! So we could get a callback called after `.close`. 

Should `event_listen` be checking `cancel_status` perhaps?

matklad commented on src/io/linux.zig:
```diff
@@ -1301,6 +1301,89 @@ pub const IO = struct {
+        completion: *Completion,
+        comptime on_event: fn (*Completion) void,
+    ) !i32 {
+        // eventfd initialized with no (zero) previous write value,
```
s/,/. ?

matklad reviewed (commented):
I feel like some sort of cancelation story is needed for Linux&Windows? (completion-based darwin should be fine perhaps)

batiati commented:
> I feel like some sort of cancelation story is needed for Linux&Windows? (completion-based darwin should be fine perhaps)

Indeed, we didn't have a well-defined cancelation story at all!

Refactored the `event` so that it does not reattach itself: it's the caller's responsibility to call `event_listen`, like the caller would `recv` a socket again after reading from it.

`Signal` is now the source of truth for shutdowns, avoiding leaving unfinished events registered.

batiati committed:
[Client] Add Events for efficient cross-thread notification.
Removes the local TCP server hack being used in `tb_client.Signal` to facilitate cross-thread notification.
Instead, it uses the most efficient mechanism for waking the IO abstraction: `eventfd` on linux, `EVFILT_USER` on macos, and `PostQueuedCompletionStatus` on Windows.

Also implements proper cancelation handling on `Signal` and `Context`.

matklad commented on src/io/test.zig:
```diff
@@ -332,6 +332,71 @@ test "timeout" {
+            // Reattaching the event.
+            self.io.event_listen(self.event, &self.event_completion, on_event);
+        }
+    }.run_test();
```
Nice pattern, will certainly steal it for myself!

matklad commented on src/io/test.zig:
```diff
@@ -332,6 +332,71 @@ test "timeout" {
+        event: IO.Event = IO.INVALID_EVENT,
+        event_completion: IO.Completion = undefined,
+
+        const delay = 100 * std.time.ns_per_ms;
```
Shorten to 5 to make the test faster? 100 ms is a lot of time!

matklad commented on src/io/linux.zig:
```diff
@@ -1367,6 +1367,76 @@ pub const IO = struct {
+        _ = self;
+
+        // eventfd initialized with no (zero) previous write value.
+        const event_fd = posix.eventfd(0, linux.EFD.CLOEXEC) catch |err| switch (err) {
```
Pre-existing, but we probably shouldn't go through `posix` here, for two reasons:

* even if libc is linked we'd rather do syscalls directly, to not mess up with errno
* we _know_ we are linux, we don't need cross-platform posix abstraction!

batiati replied:
The thing is, we'd need to handle `errno` for `linux.eventfd(...)` ourselves.
I can open a follow-up PR replacing all posix calls with native Linux ones.

matklad replied:
>The thing is, we'd need to handle errno for linux.eventfd(...) ourselves.


I don't think so? The errno is the libc thing, not the syscall thing. sys calls just return an error code as a return value as any non-insane API would do, and then libc wrapper helpfully stuffs the error code into the `errno` thread local. 

So if we use `linux.xxx` API, we don't have to touch `errno`. 

Or are you talking specifically about converting an `u32` return number to an Error value? I _think_ `std.os.linux.E.init(code)` would do that?

batiati replied:
> Or are you talking specifically about converting an u32 return number to an Error value? I think std.os.linux.E.init(code) would do that?

Yes, I mean making our `io/linux.zig` to switch over `E` instead of handling the `error` returned by the posix abstraction (not only for `eventfd`, but for all other places we use a posix call such as `write`, `open`, `close`, `fsync`, etc).

matklad commented on src/clients/c/test.zig:
```diff
@@ -244,6 +244,10 @@ test "c_client tb_status" {
 }
 
+test {
+    _ = @import("tb_client/signal.zig");
```
Include the at the top-level, in unit_tests.zig. I don't have a strong argument for one list versus
hierarchical tree of inclusions, but one list is what we do today!

batiati committed:
Code review
